File: pickerkit/src/main/java/com/dewarder/pickerkit/provider/PickerMediaImageWrapper.java
package com.dewarder.pickerkit.provider;

import com.annimon.stream.Stream;
import com.dewarder.pickerkit.model.PickerImage;
import com.dewarder.pickerkit.model.PickerMedia;

import java.util.Collection;

class PickerMediaImageWrapper implements PickerDataProvider<PickerMedia> {

    private final PickerDataProvider<PickerImage> delegate;

    private PickerMediaImageWrapper(PickerDataProvider<PickerImage> delegate) {
        this.delegate = delegate;
    }

    public static PickerDataProvider<PickerMedia> wrap(PickerDataProvider<PickerImage> provider) {
        return new PickerMediaImageWrapper(provider);
    }

    @Override
    public void request(Callback<PickerMedia> callback) {
        delegate.request(new Callback<PickerImage>() {
            @Override
            public void onNext(Collection<PickerImage> data) {
                callback.onNext(
                        Stream.of(data)
                                .map(PickerImage::getSource)
                                .map(PickerMedia.Companion::image)
                                .toList());
            }

            @Override
            public void onComplete() {
                callback.onComplete();
            }

            @Override
            public void onError(Throwable throwable) {
                callback.onError(throwable);
            }
        });
    }
}

File: pickerkit/src/main/java/com/dewarder/pickerkit/provider/PickerMediaImageWrapper.kt
package com.dewarder.pickerkit.provider

import com.dewarder.pickerkit.gallery.model.PickerImage
import com.dewarder.pickerkit.gallery.model.PickerMedia

internal class PickerMediaImageWrapper constructor(
        private val delegate: PickerDataProvider<PickerImage>
) : PickerDataProvider<PickerMedia> {

    override fun request(callback: PickerDataProvider.Callback<PickerMedia>) {
        delegate.request(object : PickerDataProvider.Callback<PickerImage> {
            override fun onNext(data: Collection<PickerImage>) {
                callback.onNext(
                        data.map(PickerImage::source).map(PickerMedia.Companion::image)
                )
            }

            override fun onComplete() {
                callback.onComplete()
            }

            override fun onError(throwable: Throwable) {
                callback.onError(throwable)
            }
        })
    }

    companion object {

        fun wrap(provider: PickerDataProvider<PickerImage>): PickerDataProvider<PickerMedia> =
                PickerMediaImageWrapper(provider)
    }
}
